binary_search_tree/tree.py

```diff
@@ -14,45 +14,139 @@ class Tree:
 
-    # Time Complexity: 
-    # Space Complexity: 
     def find(self, key):
-        pass
+        if self.root != None:
+            return self.find_helper(self.root, key)
+
+    # Time Complexity: O(n) where n is each visited node
+    # Space Complexity: O(n) - O(h) for the call stack, where h is the height of the tree
+    # Inorder  Traversal: Left - Root - Right
+    def inorder_helper(self, current, ascending_nodes):
+        if current == None:
+            return None
+        
+        self.inorder_helper(current.left, ascending_nodes)
+        ascending_nodes.append({"key": current.key, 
+                                "value": current.value})
+        self.inorder_helper(current.right, ascending_nodes)
 
-    # Time Complexity: 
-    # Space Complexity: 
     def inorder(self):
-        pass
+        if self.root == None:
+            return []
+
+        ascending_nodes = []
+        self.inorder_helper(self.root, ascending_nodes)
+
+        return ascending_nodes
+
+    # Time Complexity: O(n) where n is each visited node
+    # Space Complexity: O(n) - O(h) for the call stack, where h is the height of the tree
+    # Preoder traversal: Root - Left - Right  
+    def preorder_helper(self, current, preorder_nodes):
+        if current == None:
+            return None
+
+        preorder_nodes.append({"key": current.key, 
+                                "value": current.value})
+        self.preorder_helper(current.left, preorder_nodes)
+        self.preorder_helper(current.right, preorder_nodes)
 
-    # Time Complexity: 
-    # Space Complexity:     
     def preorder(self):
-        pass
+        if self.root == None:
+            return []
+
+        preorder_nodes = []
+        self.preorder_helper(self.root, preorder_nodes)
+
+        return preorder_nodes
+
+    # Time Complexity: O(n) where n is each visited node
+    # Space Complexity: O(n) - O(h) for the call stack, where h is the height of the tree
+    # Postorder traversal: Left - Right - Root
+    def postorder_helper(self, current, postorder_nodes):
+        if current == None:
+            return None
+        
+        self.postorder_helper(current.left, postorder_nodes)
+        self.postorder_helper(current.right, postorder_nodes)
+        postorder_nodes.append({"key": current.key, 
+                                "value": current.value})
 
-    # Time Complexity: 
-    # Space Complexity:     
     def postorder(self):
-        pass
+        if self.root == None:
+            return []
 
-    # Time Complexity: 
-    # Space Complexity:     
-    def height(self):
-        pass
+        postorder_nodes = []
+        self.postorder_helper(self.root, postorder_nodes)
+
+        return postorder_nodes
+
+    # Time Complexity: O(log n) if Balanced BST and O(n) for Unbalanced BST
+    # Space Complexity:  O(n) where n is the number of nodes in a given binary tree.
+    def height_helper(self, current, current_height):
```

Comment:
👍 The space complexity is correct if the tree could be unbalanced.  If the tree is balanced it's O( log n )